crypto/init.c

```diff
@@ -491,43 +499,90 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
         return 0;
+    }
+
+    /* Base init by default, always */
+    opts |= OPENSSL_INIT_BASE;
+
+    /* Default to reading the config file unless the caller says not to */
+    if (!(opts & OPENSSL_INIT_NO_LOAD_CONFIG)) {
+        /*
+         * If no settings are set, we add our own, which are the defaults
+         * apart from the return code ignoring flag.
+         */
+        if (settings != NULL) {
+            static const OPENSSL_INIT_SETTINGS verbose = {
+                NULL, NULL,
+                DEFAULT_CONF_MFLAGS & ~CONF_MFLAGS_IGNORE_RETURN_CODES
+            };
+            static const OPENSSL_INIT_SETTINGS noverbose = {
+                NULL, NULL,
+                DEFAULT_CONF_MFLAGS | CONF_MFLAGS_IGNORE_RETURN_CODES
+            };
+
+            switch (opts & OPENSSL_INIT_LOAD_CONFIG) {
+            case 0:              /* added default, we ignore errors */
+                settings = &noverbose;
+                break;
+            default:             /* explicitly given, don't ignore errors */
+                settings = &verbose;
+                break;
+            }
+        }
 
-    if (opts & OPENSSL_INIT_BASE_ONLY)
-        return 1;
+        opts |= OPENSSL_INIT_LOAD_CONFIG;
+    }
 
     /*
-     * Now we don't always set up exit handlers, the INIT_BASE_ONLY calls
-     * should not have the side-effect of setting up exit handlers, and
-     * therefore, this code block is below the INIT_BASE_ONLY-conditioned early
-     * return above.
+     * We want two be able to pass the settings to RUN_ONCE functions; this
+     * requires a lock, since RUN_ONCE doesn't pass an extra argument.
+     * We want to be able to have certain things happen by default (loading
+     * the config file, for example).
+     * We want to be able to call OPENSSL_init_crypto() recursively (or at
+     * least let it happen because some functions called down the line from
+     * here may as well be called directly from the application, so we want
+     * to ensure initialization in both cases).
+     *
+     * Unfortunately, combining all three may lead to nasty lock ups,
+     * because a recursive call may try to lock the same lock that was
+     * already locked.
+     *
+     * Our solution for this problem is to keep track of what things are
+     * already being initialized and should therefore not be initialized
+     * again.  We do this by saving away what options have already been
+     * chosen in an earlier call, and making opts only hold the bits that
+     * weren't already there.
+     *
+     * Note that this really makes any init option a one chance only, even
+     * if one of the initializations fail.
      */
+    {
+        CRYPTO_THREAD_write_lock(init_lock);
+        opts &= ~last_opts;      /* Mask away what's already been done */
+        last_opts |= opts;       /* Add what we're currently going to do */
+        CRYPTO_THREAD_unlock(init_lock);
```

Comment:
So we have two threads - thread 1 and 2. Both call `OPENSSL_init_crypto` with just `OPENSSL_INIT_BASE` in the `opts`. Thread 1 gets to this section first and `OPENSSL_INIT_BASE` gets added to `last_opts`. Thread 1 then gets descheduled and Thread 2 gets scheduled instead. In Thread 2 `OPENSSL_INIT_BASE` gets removed from `opts` because its already in `last_opts`. Thread 2 then continues running, skipping over all the base initialisation, and returns from OPENSSL_init_crypto and continues on into code which assumes that the base has been initialised. But Thread 1 hasn't got that far yet - its still descheduled.

Reply:
The obvious solution would be an init lock, except recursion becomes a problem...  actually the exact problem I'm trying to mitigate, 'cause there's that `init_lock` that's locked when loading the config file in current code (i.e. before this PR), and since `OPENSSL_init_crypto(0, NULL)` is called a little here and there, there's no real way to predict when such a call comes to bite us.